### ISSUE TITLE
All argument keys need to exist in file

### ISSUE THREAD

**Issue by tversteeg** [museun/markings#7]
A similar issue to #4, whenever I have a file that doesn't contain all of the template keys I get the following error:

`template key '' is missing`

**Comment by museun** [comment]
I would need an example. My use case for making this crate is that all the keys must be present. So altering that would be a deviation from the design. Originally it was made to be part of a `#[derive(template)]` type of macro that'd match enum variants up with template strings. 

I can, perhaps, come up with something like `OptionalArgs` that coincide with `Args` allowing `Template::apply` to determine whether or not missing keys should be an error.

The error you give above makes me think something internally is wrong. Can you provide a reproduction? A simple template string + the args used would be enough for me to figure out whats wrong.

**Comment by tversteeg** [comment]
So my use case is the following: I have a bunch of template files that are loaded using `include_str!()` and that I want to copy to new files while replacing all the keywords in a similar matter. The files might contain all, some or none of the keywords.

For my use case something like `OptionalArg` would be completely fine.

Here is a simple reproduction:

```rust
fn test() -> Result<()> {
    let args = Args::new()
        .with("name", &"something")
        .with("short_name", &"something else")
        .build();

    let template = Template::parse("${short_name}")?;

    println!("{}", template.apply(&args)?);

    Ok(())
}
```

**Comment by museun** [comment]
I see, I have thought about this and I think making `Template::parse` (or an aux. method) take a struct which determines the attributes of the template would solve this and allow for more flexibility in the future.

with a builder pattern like this: (option names pending, just off the top of my head.)
```rust
TemplateOpts::default()
  .optional_keys() // not all keys must be replaced
  .allow_empty() // no keys are required to be in the string
  .duplicate_keys() // duplicate keys would be replaced with the same 'arg'
  .build(); 

// TemplateOpts::default for the current behavior
let template = Template::parse(&input, &opts).unwrap();
```

It would store the options in the template itself, so `Template::apply()` would continue to have a simple interface. And still allow for my 'strict' behavior + more lax behavior w.r.t to 'well-formedness'

**Comment by tversteeg** [comment]
Sounds like a great solution for a wide variety of use cases!

**Comment by museun** [comment]
I implemented it in 1f54f11b4252e8ababf962207d2afb8dac6518bd

which should be `v0.3.0 `

This adds an Opts type which will let you configure how the parser and 'apply' works
